### PR TITLE
Buffer events to disk *before* sending, discard them on success.

### PR DESCRIPTION
**Pull request by bretthoerner**
I'm trying to handle the feedback here: https://github.com/getsentry/raven-java/issues/251#issuecomment-258597779

Basically, we should buffer events to disk (when disk buffering is enabled) *before* we send, so that uncaught exceptions can be saved before an application dies (Android is the prime example).

The PR does that, but I just remembered we wrap the connection in `AsyncConnection` at the very end of construction -- in other words even the act of buffering will be done async.

Need to think of/discuss some changes here, because I think buffering needs to be done sync, and then attempting to send/discarding-on-success can be done async.

### CODE REVIEW

**Comment by bretthoerner** [comment]
We discussed fixing the ordering problem with a beautiful double wrap such as:

`BufferedConnectionWriter(AsyncConnection*(BufferedConnectionDeleter(HttpConnection))))`

**Comment by bretthoerner** [comment]
M'kay, added an awesome writing wrapper so our connections are a nice 500 objects deep. 👻

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java [hunk 1]
```diff
@@ -139,15 +134,56 @@ public void close() throws IOException {
...
+     *         a wrapped {@link Connection}.
+     */
+    public Connection wrapConnectionWithBufferWriter(final Connection connectionToWrap) {
+        return new Connection() {
+            final Connection wrappedConnection = connectionToWrap;
+
+            @Override
+            public void send(Event event) throws ConnectionException {
+                // buffer before we attempt to send
+                buffer.add(event);
```
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: Is it OK that a failure to buffer the event will prevent the event from being attempted to be sent?

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java [hunk 2]
```diff
@@ -158,6 +194,14 @@ public void run() {
                 while (events.hasNext() && !closed) {
                     Event event = events.next();
 
+                    long now = System.currentTimeMillis();
+                    long eventTime = event.getTimestamp().getTime();
+                    long age = now - eventTime;
+                    if (age < minAgeMillis) {
+                        logger.trace("Ignoring buffered event because it only " + age + "ms old.");
+                        return;
+                    }
```
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: What's this for?

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java (hunk 1 above)
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: Good question, I guess I'll wrap it.

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java (hunk 2 above)
Comment (bretthoerner) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: So now that we store events before being sent, there is a race where you write to disk and the `Flusher` may happen to run and immediately attempt to send it while the "main" connection is also sending.

I figured it was,

1. Not 100% critical to never, ever send dupes, and not deserving of a much more elaborate system to rule out "who is sending what event this second."
2. Enough to only attempt to send buffered send events once they are at least `flushTime` (default 60s) old.

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java [hunk 3]
```diff
@@ -15,7 +15,19 @@
...
 /**
- * Connection wrapper that sends Events to an Buffer when send fails.
+ * Connection wrapper that handles storing and deleting {@link Event}s from a {@link Buffer}.
+ *
+ * This exists as a {@link Connection} implementation because the existing API (and the Java 7
+ * standard library) has no simple way of knowing whether an asynchronous request succeeded
+ * or failed. The {@link #wrapConnectionWithBufferWriter} method is used to wrap an existing
+ * Connection in a small anonymous Connection implementation that will always synchronously
+ * write the sent Event to a Buffer and then pass it to the underlying Connection (often an
+ * AsynConnection in practive). Then, an instance of the {@link BufferedConnection} is used
```
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: `s/AsynConnection/AsyncConnection/`
`s/practive/practice/`

File: raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java (hunk 2 above)
Comment (tkaemming) [inline raven/src/main/java/com/getsentry/raven/connection/BufferedConnection.java]: Makes sense, it'd be helpful to add a comment here explaining what's going on for future reference.

**Pull request merged**